### ISSUE TITLE
Default script 'bugs' with fixes.

### ISSUE THREAD

**Issue by Darkcyde13** [EDCD/EDDI#1656]
EDDI v3.5.2.

As promised, here is my list of 'bugs' I've found while going through the default scripts recently.  I've practically scrutinised every single line of code, over every single script, when rebuilding my personality, that's how I found most of these.

I went back over my change log to make sure I had got all the things I felt were 'bugs', and found a few more.  Turns out there are many problems (if I can call them that?) spread over 19 default scripts.  I've gone back over them for v3.5.2, just in case you had found any already, but none have been fixed yet.  A couple of these are now being picked up by the code colouring, showing yellow } at the ends.

**Body volcanism report**
  - Lines 36 to 39: `OneOf()` for lava/magma: Incorrect quote marks, { & }, and extra `OneOf()`.
```
     Fix: {OneOf("lava flows {Occasionally(3, 'at the surface')}",
                 "magma flows near the surface",
                 "sub-surface magma flows")}.
```

**Crew member removed**
  - Line 6: Typo, 'remove' should be 'removed'.

**Crime check station** & **Crime check system**
  - Broken `CheckPad()` function:
      Missing quote mark for "medium"
      Incorrect & missing closing brackets:
      - Line 6: Incorrect } at the end.
      - Line 13: Missing } at the end.
  This means it does not function as intended.
```
   Fix: {set CheckPad(current_station) to:
            {set largest_pad to lcase(current_station.largestpad)}
            {if largest_pad != "none":
                {if largest_pad = "large": {return true}
                |elif largest_pad = "medium":
                    {if lcase(ship.size) = "large": {return false}
                    |else: {return true}
                    }
                }
                {if lcase(ship.size) = "small": {return true} }
            }
            {return false}
        }
```

**Crime report**
  - Line 42: `systems_count` should equal the length of `systems_bounties` array, not `factions_bounties` array.

**Entered supercruise**
  - Line 7: `OneOf()` should be `Occasionally()`, to reflect the true nature of this function.

**Exploration data sold**
  - Line 1: 'systems' needs to be updated for only one system.
     `Fix: system{if len(event.systems) > 1:s}`
  - Line 4: IF checks the length of bonus data, when really it should be the value of the bonus.
     `Fix: {if event.bonus > 0:`

**Galnet news published**
  - Line 64: `= len(event.items)` should be `= len(interesting_news_title)` as per the while loops variable.
  - Line 84: Unnecessary extra quote mark (") at the end of the script.

**Launchbay report**
 - Lines 35 & 36: Double-reporting of an empty hangar.
    Fix: Delete line 36.

**Limpet purchased** & **Limpet Sold**
  - Line 2: Speech for a single limpet should be 'has', not 'have'.
     `Fix: {Occasionally(2, "{if event.amount = 1: has |else: have} been")}`

**Mission check galaxy**
  - Lines 62 to 64: Will incorrectly use "All of these are" when you only have one mission, but with multiple destinations.  It should say "It is a":
```
     Fix: {if multi = missionsCount:
              {if multi = 1: It is a |else: All of these are }
          |else:
              {multi} of these are
          }
```

**Mission completed**
  - Line 36: Typo, spelling of 'received' in the Community Goal section.
  - Line 69: Incorrect additional space in the plural IF in the commodity reward section for 'tonnes'.
     `Fix: tonne{if event.rewardCommodityAmount > 1:s}`

**Passengers**
  - Line 1: Missing comment underscore.

**Search and rescue**
  - Line 28: Unnecessary `OneOf()` in hostages section.
  - Line 36: Missing full stop to properly terminate speech.
     There are several scripts missing this, but I've not bothered to list them if that's the only problem, as most of the time it's not that noticeable.

**Ship sold**
  - Line 3: Redundant comma at the end of the initial `OneOf()` block.
  - Line 7: Missing full stop at the end to properly terminate the speech.

**System distance report** & **System report**
  - Line 4: `reportbody.name` in first IF should be `reportsystem.name`.

**System state changed**
  - System Name set up:
      Line 12: Incorrect bracket type in `|elif event.system = system.name:` section. { should be (.
     `Fix: {set system to OneOf("{P(event.system)}", "this system")}`
      In the `|else:`
      Line 15: Missing the separating comma.
      Line 16: Extra closing parenthesis ).
     `Fix: {set system to OneOf("{P(event.system)}", "system {P(event.system)}")}`

I think that's about it.  If I find any more, I'll let you know! :)

**Comment by Hoodathunk** [comment]
The 'Mission check' scripts have been substantially changed for the upcoming 'Mission Monitor Update` branch, but I will verify that the 'error' has not been duplicated in the new 'galaxy' script.

**Comment by Darkcyde13** [comment]
For some reason, something made me double check the CheckPad() function just now.  Seems my eyesight is playing tricks on me again. :P  The extra } is on line 8, not 6.

**Comment by richardbuckle** [comment]
A word to the wise, since zero generally takes the plural form in both English and the other other languages I know, I've come to prefer, e.g.: 
```
{if(systemCount = 1: system |else: systems}
```
over:
```
{if(systemCount > 1: systems |else: system}
```

If you've already filtered out the zero case then it's not, but I still recommend the habit.

**Comment by richardbuckle** [comment]
Thanks for these: will action tomorrow.

**Comment by Darkcyde13** [comment]
I agree with what you say about the plural, and normally I would have put something similar.  The reason I didn't in this case is due to the nature of the event, as you can't sell zero exploration data.

However, even though that is the case, your solution is preferable as best practice (as you say) and takes no more processing.  I'll change my script accordingly. :)

**Comment by Tkael** [comment]
Body volcanism report
- Quote marks: These are valid for the code though there's still an argument for altering them for clarity.
- { & }: I'm not sure what you mean here. Did you mean to indicate the escape characters (`/`) that are present in the current script but not in your suggested replacement?
- Extra `OneOf`: This depends on your preference as to the frequency / weighting of each response. As written, lava and magma are equally likely results (which matches my preference). With your suggested fix, 2/3 of the results will be for magma and 1/3 will be for lava. 

Crew member removed
- Typo: Agreed.

Crime check station and Crime report
- Deferred to Hoodathunk to handle in his branch.

Entered supercruise
- `OneOf` vs `Occasionally`: So `{Occasionally(2, "{ShipName()} has")} {OneOf("entered", "jumped to")} supercruise.`? Agreed. It'll be easier for a user to adjust the frequency with which the ship name is spoken with the revision.

Exploration data sold
- Line 1: Already discussed above.
- Line **3**: Agreed, the value is a decimal and not an array.

Galnet news published:
- Line 64: Agreed.
- Line 84: Agreed.

Launchbay Report
- Lines 35-36: Agreed.

Limpet purchased & Limpet Sold
- Line 2: Agreed.

Mission check galaxy and Mission completed
- Deferred to Hoodathunk to handle in his branch.

Passengers
- Line 1: Agreed, though really none of the content in this script currently adds value.

Search and rescue
- Line 28: Agreed.
- Line 36: Agreed, though except in the case of an invoked script this is not critical. (In an invoked script, the missing full stop might prevent clear separation between the invoking script and the results of the invoked script). 

Ship sold
- Line 3: Agreed.
- Line 7: Agreed, though as stated above this is not particularly critical in a script that is triggered from an event rather than invoked.

System distance report and System report
- Line 4: Referencing `reportbody.name` does not make sense here. Should be simply `{if !reportsystem.name:`.

System state changed
- Line 12: Agreed, the `{P()}` function is invoked improperly.
- Lines 15-16: Agreed.

**Comment by Darkcyde13** [comment]
Thank you for taking the time to look over them all.

For the default Body volcanism report, now that you've said that, I've looked at it again, and I do see what you mean about it being valid as it is.  I was under the impression that the options needed to be in quotes, which would also have meant that the second line needed to be in { }.  I also hadn't thought about the intention was for the first line to be spoken 50%, and the other two 25% each.  So, I apologise for my mistakes.

For Entered supercruise: yep, that's exactly how I have it in my version of the script.

For the System distance and System report, that's what I meant, reportbody.name shouldn't be there as they are system based scripts, not body ones, and 'reportybody' isn't even set anywhere.  The only reason I said to replace it with reportsystem.name (so making it  || reportsystem.name = "")  is because that's how the IF check is in other scripts, like Body report and others.  If checking for an empty string can be omitted, then can it not also be removed from those other scripts?

For the missing full stops, your assessment is pretty much the reason I mentioned it.  Not critical, but not 'correct' punctuation either.

On another note, I've been going through all my scripts again, looking for instances of plural IF statements, with the intention of changing them all over.  I've found there are many instances of the default scripts still having them like this: mission{if missions > 1:s}.  I'm thinking to maybe change them all over, and while it will add (admittedly a small) extra length to the scripts, I think it would also make script translation easier, as not all languages just add an 's' to make a word plural.

Also, while going through everything again, I've found another default script with a problem.  I've only got down to 'M' in the list so far, but the 'Mission faction state' script is missing colons ':' from the 'Outbreak', 'Retreat', and 'War' |elif: statements.  I only noticed because they were also missing from my version.  I guess I haven't looked closely enough at all of the scripts yet, if I've missed stuff! ;)
